### PR TITLE
Add support for triggers on hypertables for all triggers except INSER…

### PR DESCRIPTION
**Pull request by cevian**
…T AFTER

This adds support for all types of triggers on a hypertable except
INSERT AFTER. UPDATE and DELETE ROW triggers are automatically copied from
a hypertable onto the chunks. Therefore, any trigger defined on the
parent hypertable will apply to any row in any of the chunks as well.
STATEMENT level triggers and iNSERT triggers need not be copied in this
way.

### CODE REVIEW

File: sql/ddl_internal.sql [hunk 1]
```diff
@@ -173,6 +173,36 @@ INSERT INTO _timescaledb_catalog.hypertable_index (hypertable_id, main_schema_na
 VALUES (hypertable_id, main_schema_name, main_index_name, definition);
 $BODY$;
 
+-- Add an index to a hypertable
```
Comment (olofr) [inline sql/ddl_internal.sql]: Wrong comment ?

File: sql/chunk_trigger.sql [hunk 2]
```diff
@@ -0,0 +1,57 @@
+-- Convert a general trigger definition to a create trigger sql command for a
+-- particular table and trigger name.
```
Comment (olofr) [inline sql/chunk_trigger.sql]: I think we have a naming issue if we have both chunk_trigger.sql and chunk_triggers.sql.

File: sql/ddl_triggers.sql [hunk 3]
```diff
@@ -74,8 +79,34 @@ BEGIN
                 AND trigger_name != '_timescaledb_main_after_insert_trigger'
```
Comment (olofr) [inline sql/ddl_triggers.sql]: could we have a `get_timescale_triggers()` function and then do `where trigger_name <> ALL(get_timescale_triggers())` ?

File: sql/hypertable_trigger_triggers.sql [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    hypertable_trigger_id INTEGER
+)
+    RETURNS VOID LANGUAGE SQL VOLATILE AS
+$BODY$
+    DELETE FROM _timescaledb_catalog.chunk_trigger ci
+    WHERE ci.hypertable_trigger_id = drop_trigger_on_all_chunks.hypertable_trigger_id;
+$BODY$;
+
+
+-- Creates triggeres on chunk tables when hypertable_trigger rows created.
```
Comment (olofr) [inline sql/hypertable_trigger_triggers.sql]: typo in comment

File: sql/setup_main.sql [hunk 5]
```diff
@@ -32,6 +32,17 @@ BEGIN
     AFTER INSERT OR UPDATE OR DELETE ON _timescaledb_catalog.hypertable_index
     FOR EACH ROW EXECUTE PROCEDURE _timescaledb_internal.on_change_hypertable_index();
 
+    DROP TRIGGER IF EXISTS trigger_main_on_change_hypertable_trigger
```
Comment (olofr) [inline sql/setup_main.sql]: Could you remind we why we need `DROP TRIGGER IF EXISTS` on some but not all triggers ?

Comment (cevian) [inline sql/setup_main.sql]: All table triggers use the drop. All event triggers don't. Perhaps something to reconsider but I don't think in this PR.

File: sql/chunk_trigger.sql (hunk 2 above)
Comment (cevian) [inline sql/chunk_trigger.sql]: I deleted chunk_triggers (moved stuff to chunk.sql). Kept chunk_trigger.sql.

**Comment by cevian** [comment]
@olofr addressed all your comments. Please review again.

**Comment by erimatnor** [comment]
Another data point here is that I believe there is also another potential approach to this with my upcoming INSERT path changes. Basically, since I am doing executor/plan changes to redirect tuples from the main table to chunks, you can make the main table triggers fire when inserting on chunks. Just mentioning this to point out that there are potentially cleaner ways of doing this without new additions to our metadata schema.

**Comment by cevian** [comment]
@erimatnor  you are right, I didn't actually need the addition metadata tables. Fixed now.

**Pull request merged**